data/code/lib.c

```diff
@@ -589,3 +593,715 @@ void check_lap_ex()
+    signed int ret = 0;
+    if ((stick & 0x1) != 0)  // Right.
+    {
+        ret = -1;
+    }
+    else if ((stick & 0x2) != 0)  // Left.
+    {
+        ret = 1;
+    }
+    if (is_mirror(race_manager) == false)  // Flip if NOT mirror.
+    {
+        ret *= -1;
+    }
+    return ret;
+}
+
+// Main function for shifting sideways during bounce.
+void handle_x_adjustment(char* const this, char* const ctrl, char* const race_manager, int kart_num)
+{
+    float* const last_momentum = &s_last_momenta[kart_num];
+
+    float z_direction_vector[] = {0.0, 0.0, 0.0};
+    ObjUtility__getKartZdir(kart_num, z_direction_vector);
+
+    signed int stick_dir_id = get_stick_dir_id(ctrl, race_manager, kart_num);
+
+    float speed = 0;
+
+    if (stick_dir_id != 0)  // If holding left or right.
+    {
+        speed = add_speed(*last_momentum, stick_dir_id);
+    }
+    else
+    {
+        speed = deaccelerate_speed(*last_momentum);
+    }
+    *last_momentum = speed;  // Stores to 0x80005240 + kart_num.
+    speed *= 10.0;
+
+    z_direction_vector[0] *= speed;
+    z_direction_vector[2] *= speed;
+    add_absolute_position_vector(
+        this, z_direction_vector[0], z_direction_vector[1], z_direction_vector[2]);
+}
+
+// Resets last recorded XZ momentum before bounce liftoff;
+void reset_last_momentum(int kart_num)
+{
+    float* const last_momentum = &s_last_momenta[kart_num];
+    *last_momentum = 0;
+}
+
+void call_do_spd_ctrl_normal(char* const strat)
+{
+    KartStrat__DoSpeedCrl(strat);  // Original instruction;
+}
+
+void call_do_spd_ctrl_mod(char* const this,
+                          char* const ctrl,
+                          char* const race_manager,
+                          int kart_num)
+{
+    handle_boosts(this);
+    handle_x_adjustment(this, ctrl, race_manager, kart_num);
+    handle_y_adjustment(this, ctrl, kart_num);
+    clamp_movement_vector_descent(this);
+}
+
+void call_do_spd_ctrl(char* const this,
+                      char* const strat,
+                      char* const ctrl,
+                      char* const race_manager,
+                      int kart_num,
+                      int kart_bounce_flag)
+{
+    if (kart_bounce_flag == false)
+    {
+        call_do_spd_ctrl_normal(strat);
+    }
+    else
+    {
+        call_do_spd_ctrl_mod(this, ctrl, race_manager, kart_num);
+
+        // ASM section done here instead of its own function (has differing results?).
+        // Forcefully set registers to result of original instruction for peace of mind.
+
+        asm("lis %r4, 0x8143");
+        asm("ori %r4, %r4, 0x3a50");
+        asm("lis %r5, 0x0000");
+        asm("ori %r5, %r5, 0x0040");
+        asm("lis %r6, 0x0000");
+        asm("ori %r6, %r6, 0x0040");
+        asm("lis %r7, 0x0000");
+        asm("ori %r7, %r7, 0x0040");
+        asm("lis %r8, 0x0000");
+        asm("ori %r8, %r8, 0x0040");
+        asm("lis %r9, 0x0000");
+        asm("ori %r9, %r9, 0x000a");
```

Comment:
The `lis` instructions where the operand is `0` could be dropped, and the companion `ori` instruction would be converted to `li`.